File: rust-example/tests/capabilities/book_stay/process_tests.rs
use booking_a_stay::application_state::AppState;
use booking_a_stay::capabilities::book_stay::process::{BookStayResponse, process};
use booking_a_stay::capabilities::book_stay::request::{BookStay, Stay};
use booking_a_stay::capabilities::book_stay::result::BookingRejected;
use booking_a_stay::providers::Providers;
use chrono::NaiveDate;
use sqlx::postgres::{PgPool, PgPoolOptions};
use sqlx::{Executor, Row};
use tokio::sync::Mutex;
use uuid::Uuid;

static TEST_DATABASE_LOCK: Mutex<()> = Mutex::const_new(());

#[tokio::test]
async fn confirms_reservation_and_records_listing_unavailable_nights() {
    let _guard = TEST_DATABASE_LOCK.lock().await;
    let pool = prepare_database().await;
    let fixture = seed_bookable_listing(&pool, "eligible", "bookable").await;
    let state = app_state(pool.clone());

    let request = book_stay_request(fixture.guest_id, fixture.listing_id);
    let response = process(request, &state)
        .await
        .expect("process should not fail technically");

    let reservation_id = match response {
        BookStayResponse::Confirmed { reservation_id } => reservation_id,
        BookStayResponse::Rejected(rejection) => {
            panic!("expected confirmation, got rejection: {rejection:?}");
        }
    };

    let reservation_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM reservations")
        .fetch_one(&pool)
        .await
        .expect("reservations count should be readable");
    assert_eq!(reservation_count, 1);
    assert_eq!(
        stored_reservation_id(&pool).await,
        reservation_id,
        "RPU-generated reservation id should be recorded"
    );

    let unavailable_nights = unavailable_nights(&pool, fixture.listing_id).await;
    assert_eq!(
        unavailable_nights,
        vec![date(2026, 8, 1), date(2026, 8, 2), date(2026, 8, 3)]
    );
}

#[tokio::test]
async fn does_not_occupy_the_check_out_date() {
    let _guard = TEST_DATABASE_LOCK.lock().await;
    let pool = prepare_database().await;
    let fixture = seed_bookable_listing(&pool, "eligible", "bookable").await;
    let state = app_state(pool.clone());

    let request = book_stay_request(fixture.guest_id, fixture.listing_id);
    process(request, &state)
        .await
        .expect("process should not fail technically");

    let check_out_count: i64 =
        sqlx::query_scalar("SELECT COUNT(*) FROM listing_unavailable_nights WHERE night = $1")
            .bind(date(2026, 8, 4))
            .fetch_one(&pool)
            .await
            .expect("check-out count should be readable");

    assert_eq!(check_out_count, 0);
}

#[tokio::test]
async fn rejects_when_guest_is_blocked() {
    let _guard = TEST_DATABASE_LOCK.lock().await;
    let pool = prepare_database().await;
    let fixture = seed_bookable_listing(&pool, "blocked", "bookable").await;
    let state = app_state(pool.clone());

    let request = book_stay_request(fixture.guest_id, fixture.listing_id);
    let response = process(request, &state)
        .await
        .expect("process should not fail technically");

    assert_eq!(
        response,
        BookStayResponse::Rejected(BookingRejected::GuestBlocked)
    );
    assert_no_reservation_or_unavailable_nights(&pool).await;
}

#[tokio::test]
async fn rejects_when_listing_is_disabled() {
    let _guard = TEST_DATABASE_LOCK.lock().await;
    let pool = prepare_database().await;
    let fixture = seed_bookable_listing(&pool, "eligible", "disabled").await;
    let state = app_state(pool.clone());

    let request = book_stay_request(fixture.guest_id, fixture.listing_id);
    let response = process(request, &state)
        .await
        .expect("process should not fail technically");

    assert_eq!(
        response,
        BookStayResponse::Rejected(BookingRejected::ListingDisabled)
    );
    assert_no_reservation_or_unavailable_nights(&pool).await;
}

#[tokio::test]
async fn rejects_when_requested_night_is_already_unavailable() {
    let _guard = TEST_DATABASE_LOCK.lock().await;
    let pool = prepare_database().await;
    let fixture = seed_bookable_listing(&pool, "eligible", "bookable").await;
    let state = app_state(pool.clone());

    sqlx::query(
        r#"
        INSERT INTO listing_unavailable_nights (
            listing_id,
            night,
            unavailability_type,
            reason
        )
        VALUES ($1, $2, 'host_block', 'maintenance')
        "#,
    )
    .bind(fixture.listing_id)
    .bind(date(2026, 8, 2))
    .execute(&pool)
    .await
    .expect("host block should be inserted");

    let request = book_stay_request(fixture.guest_id, fixture.listing_id);
    let response = process(request, &state)
        .await
        .expect("process should not fail technically");

    assert_eq!(
        response,
        BookStayResponse::Rejected(BookingRejected::ListingUnavailable)
    );

    let reservation_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM reservations")
        .fetch_one(&pool)
        .await
        .expect("reservations count should be readable");
    assert_eq!(reservation_count, 0);
}

#[tokio::test]
async fn two_overlapping_booking_requests_cannot_both_confirm() {
    let _guard = TEST_DATABASE_LOCK.lock().await;
    let pool = prepare_database().await;
    let fixture = seed_bookable_listing(&pool, "eligible", "bookable").await;
    let state = app_state(pool.clone());

    let first_request = book_stay_request(fixture.guest_id, fixture.listing_id);
    let second_request = book_stay_request(fixture.guest_id, fixture.listing_id);

    let first_state = state.clone();
    let second_state = state.clone();

    let (first, second) = tokio::join!(
        process(first_request, &first_state),
        process(second_request, &second_state)
    );

    let responses = vec![
        first.expect("first process should not fail technically"),
        second.expect("second process should not fail technically"),
    ];

    let confirmed_count = responses
        .iter()
        .filter(|response| matches!(response, BookStayResponse::Confirmed { .. }))
        .count();
    let unavailable_rejection_count = responses
        .iter()
        .filter(|response| {
            matches!(
                response,
                BookStayResponse::Rejected(BookingRejected::ListingUnavailable)
            )
        })
        .count();

    assert_eq!(confirmed_count, 1);
    assert_eq!(unavailable_rejection_count, 1);

    let unavailable_row_count: i64 =
        sqlx::query_scalar("SELECT COUNT(*) FROM listing_unavailable_nights")
            .fetch_one(&pool)
            .await
            .expect("unavailable night count should be readable");
    let distinct_unavailable_row_count: i64 = sqlx::query_scalar(
        "SELECT COUNT(DISTINCT (listing_id, night)) FROM listing_unavailable_nights",
    )
    .fetch_one(&pool)
    .await
    .expect("distinct unavailable night count should be readable");

    assert_eq!(unavailable_row_count, 3);
    assert_eq!(distinct_unavailable_row_count, 3);
}

async fn prepare_database() -> PgPool {
    let pool = PgPoolOptions::new()
        .max_connections(10)
        .connect(&database_url())
        .await
        .expect("test database should be reachable");

    pool.execute(
        r#"
        DROP SCHEMA public CASCADE;
        CREATE SCHEMA public;
        GRANT ALL ON SCHEMA public TO PUBLIC;
        "#,
    )
    .await
    .expect("test database schema should reset");

    sqlx::migrate!("./migrations")
        .run(&pool)
        .await
        .expect("migrations should run");

    pool
}

fn database_url() -> String {
    std::env::var("DATABASE_URL")
        .unwrap_or_else(|_| "postgres://postgres:postgres@localhost:5432/booking_a_stay".to_owned())
}

#[derive(Debug, Clone, Copy)]
struct BookingFixture {
    guest_id: Uuid,
    listing_id: Uuid,
}

async fn seed_bookable_listing(
    pool: &PgPool,
    guest_eligibility: &str,
    listing_status: &str,
) -> BookingFixture {
    let fixture = BookingFixture {
        guest_id: uuid(1),
        listing_id: uuid(2),
    };

    sqlx::query(
        r#"
        INSERT INTO guests (id, email, full_name, booking_eligibility)
        VALUES ($1, 'guest@example.test', 'Test Guest', $2)
        "#,
    )
    .bind(fixture.guest_id)
    .bind(guest_eligibility)
    .execute(pool)
    .await
    .expect("guest should be inserted");

    sqlx::query(
        r#"
        INSERT INTO listings (
            id,
            host_id,
            title,
            max_guests,
            min_nights,
            max_nights,
            booking_status
        )
        VALUES ($1, $2, 'Test Listing', 4, 2, 7, $3)
        "#,
    )
    .bind(fixture.listing_id)
    .bind(uuid(4))
    .bind(listing_status)
    .execute(pool)
    .await
    .expect("listing should be inserted");

    fixture
}

fn book_stay_request(guest_id: Uuid, listing_id: Uuid) -> BookStay {
    BookStay {
        guest_id,
        listing_id,
        stay: Stay {
            check_in: date(2026, 8, 1),
            check_out: date(2026, 8, 4),
        },
        guest_count: 2,
    }
}

fn app_state(pool: PgPool) -> AppState {
    AppState::new(pool, Providers::new())
}

async fn stored_reservation_id(pool: &PgPool) -> Uuid {
    sqlx::query_scalar("SELECT id FROM reservations")
        .fetch_one(pool)
        .await
        .expect("reservation id should be readable")
}

async fn unavailable_nights(pool: &PgPool, listing_id: Uuid) -> Vec<NaiveDate> {
    sqlx::query(
        r#"
        SELECT night
        FROM listing_unavailable_nights
        WHERE listing_id = $1
        ORDER BY night
        "#,
    )
    .bind(listing_id)
    .fetch_all(pool)
    .await
    .expect("unavailable nights should be readable")
    .into_iter()
    .map(|row| row.get("night"))
    .collect()
}

async fn assert_no_reservation_or_unavailable_nights(pool: &PgPool) {
    let reservation_count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM reservations")
        .fetch_one(pool)
        .await
        .expect("reservations count should be readable");
    let unavailable_count: i64 =
        sqlx::query_scalar("SELECT COUNT(*) FROM listing_unavailable_nights")
            .fetch_one(pool)
            .await
            .expect("unavailable nights count should be readable");

    assert_eq!(reservation_count, 0);
    assert_eq!(unavailable_count, 0);
}

fn date(year: i32, month: u32, day: u32) -> NaiveDate {
    NaiveDate::from_ymd_opt(year, month, day).expect("test date should be valid")
}

fn uuid(value: u128) -> Uuid {
    Uuid::from_u128(value)
}

File: csharp-example/tests/BookingAStay.Tests/Capabilities/BookStay/ProcessTests.cs
using BookingAStay.ApplicationState.Db;
using BookingAStay.Capabilities.BookStay;
using BookingAStay.Providers;
using Npgsql;

namespace BookingAStay.Tests.Capabilities.BookStay;

[Collection(PostgreSqlCollection.Name)]
public sealed class ProcessTests
{
    private readonly PostgreSqlFixture _fixture;

    public ProcessTests(PostgreSqlFixture fixture)
    {
        _fixture = fixture;
    }

    [Fact]
    public async Task ConfirmsReservationAndRecordsListingUnavailableNights()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "bookable");

        var request = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var response = await Processor.ProcessBookStay(request, dataSource, new ProviderBundle());

        var confirmed = Assert.IsType<BookStayResponse.Confirmed>(response);

        Assert.Equal(1, await Count(dataSource, "reservations"));
        Assert.Equal(confirmed.ReservationId, await StoredReservationId(dataSource));
        Assert.Equal(
            [Date(2026, 8, 1), Date(2026, 8, 2), Date(2026, 8, 3)],
            await UnavailableNights(dataSource, fixture.ListingId));
    }

    [Fact]
    public async Task RecordsBookingTermsUsedAtConfirmation()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedListingWithTerms(dataSource, "eligible", "bookable", 5, 2, 7);

        var request = BookStayRequest(fixture.GuestId, fixture.ListingId) with { GuestCount = 5 };
        var response = await Processor.ProcessBookStay(request, dataSource, new ProviderBundle());

        Assert.IsType<BookStayResponse.Confirmed>(response);
        Assert.Equal(5, await StoredMaxGuestsAtConfirmation(dataSource));

        await using (var connection = await dataSource.OpenConnectionAsync())
        await using (var command = new NpgsqlCommand("UPDATE listings SET max_guests = 4 WHERE id = @listing_id", connection))
        {
            command.Parameters.AddWithValue("listing_id", fixture.ListingId);
            await command.ExecuteNonQueryAsync();
        }

        Assert.Equal(5, await StoredMaxGuestsAtConfirmation(dataSource));
    }

    [Fact]
    public async Task DoesNotOccupyTheCheckOutDate()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "bookable");

        var request = BookStayRequest(fixture.GuestId, fixture.ListingId);
        await Processor.ProcessBookStay(request, dataSource, new ProviderBundle());

        await using var connection = await dataSource.OpenConnectionAsync();
        await using var command = new NpgsqlCommand(
            "SELECT COUNT(*) FROM listing_unavailable_nights WHERE night = @night",
            connection);
        command.Parameters.AddWithValue("night", Date(2026, 8, 4));

        Assert.Equal(0, (long)(await command.ExecuteScalarAsync() ?? 0L));
    }

    [Fact]
    public async Task RejectsWhenGuestIsBlocked()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "blocked", "bookable");

        var request = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var response = await Processor.ProcessBookStay(request, dataSource, new ProviderBundle());

        Assert.Equal(new BookStayResponse.Rejected(BookingRejection.GuestBlocked), response);
        await AssertNoReservationOrUnavailableNights(dataSource);
    }

    [Fact]
    public async Task RejectsWhenListingIsDisabled()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "disabled");

        var request = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var response = await Processor.ProcessBookStay(request, dataSource, new ProviderBundle());

        Assert.Equal(new BookStayResponse.Rejected(BookingRejection.ListingDisabled), response);
        await AssertNoReservationOrUnavailableNights(dataSource);
    }

    [Fact]
    public async Task RejectsWhenRequestedNightIsAlreadyUnavailable()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "bookable");

        await using (var connection = await dataSource.OpenConnectionAsync())
        await using (var command = new NpgsqlCommand(
            """
            INSERT INTO listing_unavailable_nights (
                listing_id,
                night,
                unavailability_type,
                reason
            )
            VALUES (@listing_id, @night, 'host_block', 'maintenance')
            """,
            connection))
        {
            command.Parameters.AddWithValue("listing_id", fixture.ListingId);
            command.Parameters.AddWithValue("night", Date(2026, 8, 2));
            await command.ExecuteNonQueryAsync();
        }

        var request = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var response = await Processor.ProcessBookStay(request, dataSource, new ProviderBundle());

        Assert.Equal(new BookStayResponse.Rejected(BookingRejection.ListingUnavailable), response);
        Assert.Equal(0, await Count(dataSource, "reservations"));
    }

    [Fact]
    public async Task TwoOverlappingBookingRequestsCannotBothConfirm()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "bookable");

        var firstRequest = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var secondRequest = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var providers = new ProviderBundle();

        var responses = await Task.WhenAll(
            Processor.ProcessBookStay(firstRequest, dataSource, providers),
            Processor.ProcessBookStay(secondRequest, dataSource, providers));

        var confirmedCount = responses.Count(response => response is BookStayResponse.Confirmed);
        var unavailableRejectionCount = responses.Count(response =>
            response is BookStayResponse.Rejected { Reason: BookingRejection.ListingUnavailable });

        Assert.Equal(1, confirmedCount);
        Assert.Equal(1, unavailableRejectionCount);
        Assert.Equal(3, await Count(dataSource, "listing_unavailable_nights"));
        Assert.Equal(3, await DistinctUnavailableNightCount(dataSource));
    }

    [Fact]
    public async Task TwoGuestsWithDisjointStaysBothConfirmConcurrently()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "bookable");
        var secondGuestId = await SeedSecondGuest(dataSource);
        var providers = new ProviderBundle();

        var firstRequest = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var secondRequest = BookStayRequest(secondGuestId, fixture.ListingId) with
        {
            Stay = new Stay(Date(2026, 8, 10), Date(2026, 8, 13)),
        };

        var responses = await Task.WhenAll(
            Processor.ProcessBookStay(firstRequest, dataSource, providers),
            Processor.ProcessBookStay(secondRequest, dataSource, providers));

        Assert.IsType<BookStayResponse.Confirmed>(responses[0]);
        Assert.IsType<BookStayResponse.Confirmed>(responses[1]);
        Assert.Equal(2, await Count(dataSource, "reservations"));
        Assert.Equal(6, await Count(dataSource, "listing_unavailable_nights"));
        Assert.Equal(
            [
                Date(2026, 8, 1),
                Date(2026, 8, 2),
                Date(2026, 8, 3),
                Date(2026, 8, 10),
                Date(2026, 8, 11),
                Date(2026, 8, 12),
            ],
            await UnavailableNights(dataSource, fixture.ListingId));
    }

    [Fact]
    public async Task BackToBackStaysBothConfirmConcurrently()
    {
        await using var dataSource = await PrepareDatabase();
        var fixture = await SeedBookableListing(dataSource, "eligible", "bookable");
        var secondGuestId = await SeedSecondGuest(dataSource);
        var providers = new ProviderBundle();

        var firstRequest = BookStayRequest(fixture.GuestId, fixture.ListingId);
        var secondRequest = BookStayRequest(secondGuestId, fixture.ListingId) with
        {
            Stay = new Stay(Date(2026, 8, 4), Date(2026, 8, 7)),
        };

        var responses = await Task.WhenAll(
            Processor.ProcessBookStay(firstRequest, dataSource, providers),
            Processor.ProcessBookStay(secondRequest, dataSource, providers));

        Assert.IsType<BookStayResponse.Confirmed>(responses[0]);
        Assert.IsType<BookStayResponse.Confirmed>(responses[1]);
        Assert.Equal(6, await Count(dataSource, "listing_unavailable_nights"));
        Assert.Equal(
            [
                Date(2026, 8, 1),
                Date(2026, 8, 2),
                Date(2026, 8, 3),
                Date(2026, 8, 4),
                Date(2026, 8, 5),
                Date(2026, 8, 6),
            ],
            await UnavailableNights(dataSource, fixture.ListingId));
    }

    private async Task<NpgsqlDataSource> PrepareDatabase()
    {
        var dataSource = NpgsqlDataSource.Create(_fixture.ConnectionString);

        await using (var connection = await dataSource.OpenConnectionAsync())
        await using (var command = new NpgsqlCommand(
            """
            DROP SCHEMA public CASCADE;
            CREATE SCHEMA public;
            GRANT ALL ON SCHEMA public TO PUBLIC;
            """,
            connection))
        {
            await command.ExecuteNonQueryAsync();
        }

        Migrations.Apply(_fixture.ConnectionString);
        return dataSource;
    }

    private static async Task<BookingFixture> SeedBookableListing(
        NpgsqlDataSource dataSource,
        string guestEligibility,
        string listingStatus)
    {
        return await SeedListingWithTerms(dataSource, guestEligibility, listingStatus, 4, 2, 7);
    }

    private static async Task<BookingFixture> SeedListingWithTerms(
        NpgsqlDataSource dataSource,
        string guestEligibility,
        string listingStatus,
        int maxGuests,
        int minNights,
        int? maxNights)
    {
        var fixture = new BookingFixture(GuidFromByte(1), GuidFromByte(2));

        await using var connection = await dataSource.OpenConnectionAsync();

        await using (var guestCommand = new NpgsqlCommand(
            """
            INSERT INTO guests (id, email, full_name, booking_eligibility)
            VALUES (@id, 'guest@example.test', 'Test Guest', @booking_eligibility)
            """,
            connection))
        {
            guestCommand.Parameters.AddWithValue("id", fixture.GuestId);
            guestCommand.Parameters.AddWithValue("booking_eligibility", guestEligibility);
            await guestCommand.ExecuteNonQueryAsync();
        }

        await using (var listingCommand = new NpgsqlCommand(
            """
            INSERT INTO listings (
                id,
                host_id,
                title,
                max_guests,
                min_nights,
                max_nights,
                booking_status
            )
            VALUES (@id, @host_id, 'Test Listing', @max_guests, @min_nights, @max_nights, @booking_status)
            """,
            connection))
        {
            listingCommand.Parameters.AddWithValue("id", fixture.ListingId);
            listingCommand.Parameters.AddWithValue("host_id", GuidFromByte(4));
            listingCommand.Parameters.AddWithValue("max_guests", maxGuests);
            listingCommand.Parameters.AddWithValue("min_nights", minNights);
            listingCommand.Parameters.AddWithValue("max_nights", (object?)maxNights ?? DBNull.Value);
            listingCommand.Parameters.AddWithValue("booking_status", listingStatus);
            await listingCommand.ExecuteNonQueryAsync();
        }

        return fixture;
    }

    private static async Task<Guid> SeedSecondGuest(NpgsqlDataSource dataSource)
    {
        var guestId = GuidFromByte(3);

        await using var connection = await dataSource.OpenConnectionAsync();
        await using var command = new NpgsqlCommand(
            """
            INSERT INTO guests (id, email, full_name, booking_eligibility)
            VALUES (@id, 'second-guest@example.test', 'Second Test Guest', 'eligible')
            """,
            connection);
        command.Parameters.AddWithValue("id", guestId);
        await command.ExecuteNonQueryAsync();

        return guestId;
    }

    private static BookStayRequest BookStayRequest(Guid guestId, Guid listingId)
    {
        return new BookStayRequest(
            guestId,
            listingId,
            new Stay(Date(2026, 8, 1), Date(2026, 8, 4)),
            2);
    }

    private static async Task<Guid> StoredReservationId(NpgsqlDataSource dataSource)
    {
        return await Scalar<Guid>(dataSource, "SELECT id FROM reservations");
    }

    private static async Task<int> StoredMaxGuestsAtConfirmation(NpgsqlDataSource dataSource)
    {
        return await Scalar<int>(dataSource, "SELECT max_guests_at_confirmation FROM reservations");
    }

    private static async Task<IReadOnlyList<DateOnly>> UnavailableNights(NpgsqlDataSource dataSource, Guid listingId)
    {
        await using var connection = await dataSource.OpenConnectionAsync();
        await using var command = new NpgsqlCommand(
            """
            SELECT night
            FROM listing_unavailable_nights
            WHERE listing_id = @listing_id
            ORDER BY night
            """,
            connection);
        command.Parameters.AddWithValue("listing_id", listingId);
        await using var reader = await command.ExecuteReaderAsync();

        var nights = new List<DateOnly>();
        while (await reader.ReadAsync())
        {
            nights.Add(reader.GetFieldValue<DateOnly>(0));
        }

        return nights;
    }

    private static async Task AssertNoReservationOrUnavailableNights(NpgsqlDataSource dataSource)
    {
        Assert.Equal(0, await Count(dataSource, "reservations"));
        Assert.Equal(0, await Count(dataSource, "listing_unavailable_nights"));
    }

    private static async Task<long> Count(NpgsqlDataSource dataSource, string tableName)
    {
        return await Scalar<long>(dataSource, $"SELECT COUNT(*) FROM {tableName}");
    }

    private static async Task<long> DistinctUnavailableNightCount(NpgsqlDataSource dataSource)
    {
        return await Scalar<long>(
            dataSource,
            "SELECT COUNT(DISTINCT (listing_id, night)) FROM listing_unavailable_nights");
    }

    private static async Task<T> Scalar<T>(NpgsqlDataSource dataSource, string sql)
    {
        await using var connection = await dataSource.OpenConnectionAsync();
        await using var command = new NpgsqlCommand(sql, connection);
        return (T)(await command.ExecuteScalarAsync() ?? throw new InvalidOperationException("Expected scalar value."));
    }

    private static DateOnly Date(int year, int month, int day)
    {
        return new DateOnly(year, month, day);
    }

    private static Guid GuidFromByte(uint value)
    {
        Span<byte> bytes = stackalloc byte[16];
        bytes[15] = (byte)value;
        return new Guid(bytes, bigEndian: true);
    }

    private sealed record BookingFixture(Guid GuestId, Guid ListingId);
}
